### PR TITLE
zed-editor: fix darwin

### LINKED ISSUE #320084: Build failure: zed-editor
### Steps To Reproduce

Zed-editor fails to build for MaxOS Intel using `NIXPKGS_ALLOW_BROKEN=1 nix-shell -p zed-editor -I nixpkgs=channel:nixos-unstable`

### Build log

```
  ...
  cargo:rerun-if-changed=/nix/store/xz8j1amvgwas4nbfqgpmcz147pkd391s-apple-framework-VideoToolbox/Library/Frameworks/VideoToolbox.framework/Headers/VTBase.h
  cargo:rerun-if-changed=/nix/store/wahwaqbn6f57ybai7akc5920vrln9s2j-apple-framework-CoreFoundation/Library/Frameworks/CoreFoundation.framework/Headers/CoreFoundation.h
  cargo:rerun-if-changed=/nix/store/xz8j1amvgwas4nbfqgpmcz147pkd391s-apple-framework-VideoToolbox/Library/Frameworks/VideoToolbox.framework/Headers/VTErrors.h

  --- stderr
  thread 'main' panicked at /private/tmp/nix-build-zed-0.137.6.drv-0/cargo-vendor-dir/bindgen-0.65.1/ir/ty.rs:1074:22:
  Not able to resolve array element?: Continue
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: builder for '/nix/store/dv6kd8ljw54aw4j728cy97mqg0jxlv6y-zed-0.137.6.drv' failed with exit code 101;
       last 10 log lines:
       >   cargo:rerun-if-changed=/nix/store/a5swqw43q2yahpawgwbrsgildzqni5n1-apple-framework-CoreMedia/Library/Frameworks/CoreMedia.framework/Headers/CMBase.h
       >   cargo:rerun-if-changed=/nix/store/xz8j1amvgwas4nbfqgpmcz147pkd391s-apple-framework-VideoToolbox/Library/Frameworks/VideoToolbox.framework/Headers/VTBase.h
       >   cargo:rerun-if-changed=/nix/store/wahwaqbn6f57ybai7akc5920vrln9s2j-apple-framework-CoreFoundation/Library/Frameworks/CoreFoundation.framework/Headers/CoreFoundation.h
       >   cargo:rerun-if-changed=/nix/store/xz8j1amvgwas4nbfqgpmcz147pkd391s-apple-framework-VideoToolbox/Library/Frameworks/VideoToolbox.framework/Headers/VTErrors.h
       >
       >   --- stderr
       >   thread 'main' panicked at /private/tmp/nix-build-zed-0.137.6.drv-0/cargo-vendor-dir/bindgen-0.65.1/ir/ty.rs:1074:22:
       >   Not able to resolve array element?: Continue
       >   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run 'nix log /nix/store/dv6kd8ljw54aw4j728cy97mqg0jxlv6y-zed-0.137.6.drv'.
```

### Notify maintainers

@GaetanLepage 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 23.5.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.18.2`
 - channels(ari): `"darwin"`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by niklaskorz**
## Description of changes

Fixes #320084.

~~Once https://github.com/zed-industries/zed/pull/13343 lands, building Zed on macOS becomes much easier as we do not have to deal with building a separate Swift library that Zed depends on anymore.~~

As https://github.com/zed-industries/zed/pull/13343 seems to need much longer to implement than I originally anticipated, I updated this PR to disable livekit until this change lands, so we can already go ahead and release darwin support without it, using the same internal fallbacks the Linux version is using.

As this PR is making use of the new pattern for using macOS SDKs, it's targetting staging-next, so merging this PR will have to wait until staging-next has been merged into master (and this PR has been retargetted at master). 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by niklaskorz** [comment]
@reckenrode FYI, using `overrideSDK stdenv "12.3"` here had two issues (which is why this PR does not use it):

1. `ScreenCaptureKit` is not available as attribute on the unversioned SDK, so it can only be accessed through the versioned 12.3 SDK
2. The 12.3 SDK version of the IOKit framework currently fails to build:
  ```
>   /nix/store/izq09jq3hyh53y2ha8x3hn7bfhivz6w5-apple-framework-IOKit-12.3/Library/Frameworks/IOKit.framework/Headers/IOTypes.h:81:49: error: expected ';' after top level declarator
>   /nix/store/izq09jq3hyh53y2ha8x3hn7bfhivz6w5-apple-framework-IOKit-12.3/Library/Frameworks/IOKit.framework/Headers/OSMessageNotification.h:120:53: error: expected ';' after top level declarator
>   thread 'main' panicked at crates/media/build.rs:34:10:
>   unable to generate bindings: ClangDiagnostic("/nix/store/izq09jq3hyh53y2ha8x3hn7bfhivz6w5-apple-framework-IOKit-12.3/Library/Frameworks/IOKit.framework/Headers/IOTypes.h:81:49: error: expected ';' after top level declarator\n/nix/store/izq09jq3hyh53y2ha8x3hn7bfhivz6w5-apple-framework-IOKit-12.3/Library/Frameworks/IOKit.framework/Headers/OSMessageNotification.h:120:53: error: expected ';' after top level declarator\n")
  ```

**Comment by reckenrode** [comment]
> 1. `ScreenCaptureKit` is not available as attribute on the unversioned SDK, so it can only be accessed through the versioned 12.3 SDK

Unfortunately, this is a limitation of the way the SDKs are built. Adding libs.simd was easy, but adding all the different frameworks would be more work. The workaround (as you found) is to use the versioned reference. 

Longer term, the SDK work I am doing (targeting 24.11) will eliminate individual frameworks. The stdenv will have an `SDKROOT` (plus appropriate variants for cross targets) pointing to the SDK. SDK mismatches will be a thing of the past.

> 2. The 12.3 SDK version of the IOKit framework currently fails to build:
> 
>   ```
> 
> >   /nix/store/izq09jq3hyh53y2ha8x3hn7bfhivz6w5-apple-framework-IOKit-12.3/Library/Frameworks/IOKit.framework/Headers/IOTypes.h:81:49: error: expected ';' after top level declarator
> 
> >   /nix/store/izq09jq3hyh53y2ha8x3hn7bfhivz6w5-apple-framework-IOKit-12.3/Library/Frameworks/IOKit.framework/Headers/OSMessageNotification.h:120:53: error: expected ';' after top level declarator
> 
> >   thread 'main' panicked at crates/media/build.rs:34:10:
> 
> >   unable to generate bindings: ClangDiagnostic("/nix/store/izq09jq3hyh53y2ha8x3hn7bfhivz6w5-apple-framework-IOKit-12.3/Library/Frameworks/IOKit.framework/Headers/IOTypes.h:81:49: error: expected ';' after top level declarator\n/nix/store/izq09jq3hyh53y2ha8x3hn7bfhivz6w5-apple-framework-IOKit-12.3/Library/Frameworks/IOKit.framework/Headers/OSMessageNotification.h:120:53: error: expected ';' after top level declarator\n")

Overriding the SDK with Rust is difficult because of the way `rustPlatform` is defined. It captures a bunch of different references in the top-level `rust`, which are used in the setup hooks in `rustPlatform`. `overrideSDK` is blind to hooks that capture dependencies, so it can’t override the Libsystem used by their clangs.

However, a few questions: Is this x86_64-darwin? If so, can you try staging-next? The ld64 update introduces a replacement set of availability headers in the 10.12 SDK that should be more compatible when mixed. Are you able to confirm which clang/Libsystem is being used in the failed invocation (`<path to>/clang -v -E -`)?


> Only picking ScreenCaptureKit and VideoToolbox (because of a missing symbol in the 11.0 version of VideoToolbox) from 12.3 worked fine though, so I left all other frameworks at the default SDK version.

That seems fine. Ideally, that wouldn’t be necessary. You should be able to declare the SDK you need, and everything in your build will use it. I have a PoC that does that, but it’s still very early. It doesn’t even build a simple derivation yet.

**Comment by reckenrode** [comment]
Oh, okay. Thanks for checking! It appears not even modern SDKs can be mixed. That’s the problem with IOKit. It’s not broken as long as clang is using Libsystem from the same SDK (12.3 in this case), but clang is using the 11.0 SDK (due to [this][1]).

[1]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/lib/default.nix

**Comment by vcunat** [comment]
(former `staging-next` is now in `master`)

**Comment by niklaskorz** [comment]
Updated to https://github.com/zed-industries/zed/pull/13343/commits/d47ccdb83b298930c34e608d30565f86e5938992, currently fails to build:

```
install: skipping file '/dev/fd/63', as it was replaced while being copied
ln: failed to create symbolic link '/nix/store/m0mhskjpmdysjr5fqx555jz1pbldhpfq-cargo-vendor-dir/cpal-0.15.3/hrvr4nc3q9yk3sjcal4cpghzi4mq4r41-cpal-0.15.3': Permission denied
```

**Comment by niklaskorz** [comment]
Fixed build by applying the patch from https://github.com/zed-industries/zed/pull/13343#issuecomment-2294797663. Will hopefully be fixed upstream so we don't have to patch dependencies in the long run.

**Comment by niklaskorz** [comment]
@GaetanLepage @jansol This PR should also enable collaboration features on Linux, though I haven't yet tested if it builds there. The upstream PR still isn't entirely finished so this stays a draft for now, but I think it might already be a good time to gather some feedback if you have any.

**Comment by tymscar** [comment]
If this has been closed, should we still have the package marked as broken @niklaskorz ?

**Comment by niklaskorz** [comment]
> If this has been closed, should we still have the package marked as broken @niklaskorz ?

It's only temporarily closed as I'm updating it to use the new Apple SDK pattern.

File: pkgs/by-name/ze/zed-editor/package.nix [hunk 1]
```diff
@@ -93,14 +95,23 @@ rustPlatform.buildRustPackage rec {
     repo = "zed";
     rev = "refs/tags/v${version}";
     hash = "sha256-xtSdlzj1AxhJN4aXLJ+Oy51LX4QduLwcuCfK42kthvE=";
-    fetchSubmodules = true;
```
Comment (niklaskorz) [inline pkgs/by-name/ze/zed-editor/package.nix:96]: This seems to have become obsolete for a while now, so I disabled fetching submodules, but correct me if I'm wrong.

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 329653`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>zed-editor</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>zed-editor</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>zed-editor</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>zed-editor</li>
  </ul>
</details>

**Comment by niklaskorz** [comment]
> ## `nixpkgs-review` result
> 
> Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).
> 
> Command: `nixpkgs-review pr 329653`
> ### `x86_64-linux`
> ✅ 1 package built:
> 
> 
> ### `aarch64-linux`
> ✅ 1 package built:
> 
> 
> ### `x86_64-darwin`
> ❌ 1 package failed to build:
> 
> 
> ### `aarch64-darwin`
> ❌ 1 package failed to build:

Could you provide any logs for the failing builds? As it builds fine on aarch64-darwin here. I'll try x86_64-darwin in a moment.

**Comment by GaetanLepage** [comment]
> Could you provide any logs for the failing builds? As it builds fine on aarch64-darwin here. I'll try x86_64-darwin in a moment.

Actually I naively ran the tool and didn't notice that the PR was targetting staging.
I think it failed on darwin because cargo wasn't building.

**Comment by niklaskorz** [comment]
> > Could you provide any logs for the failing builds? As it builds fine on aarch64-darwin here. I'll try x86_64-darwin in a moment.
> 
> Actually I naively ran the tool and didn't notice that the PR was targetting staging. I think it failed on darwin because cargo wasn't building.

I see! Yeah, it takes a good while because Rust and Cargo 1.82 are not cached yet on staging-next, but both should build fine :)

**Comment by GaetanLepage** [comment]
Good job on this @niklaskorz !
Do you want to merge this against `staging-next` or wait that it lands on `master` first and then merge it on `master` ?

File: pkgs/by-name/ze/zed-editor/package.nix [hunk 2]
```diff
@@ -191,6 +195,9 @@ rustPlatform.buildRustPackage rec {
     ZED_UPDATE_EXPLANATION = "zed has been installed using nix. Auto-updates have thus been disabled.";
```
Comment (jansol) [inline pkgs/by-name/ze/zed-editor/package.nix:195]:
```suggestion
    ZED_UPDATE_EXPLANATION = "Zed has been installed using nix. Auto-updates have thus been disabled.";
```

Comment (niklaskorz) [reply pkgs/by-name/ze/zed-editor/package.nix:195]: Unrelated to this PR, but I agree that looks nicer, will add it as a separate commit :)

**Comment by niklaskorz** [comment]
> Good job on this @niklaskorz ! Do you want to merge this against `staging-next` or wait that it lands on `master` first and then merge it on `master` ?

I am fine with both. On the "Nix on macOS" Matrix room I asked what the usual procedure was and @khaneliman answered:

> If there's a dependency on something in staging/staging-next I don't think it's necessarily wrong to target them if you don't want to wait until it propagates down. I think there's a chance stuff gets ripped out or reverted in those branches, though so it might make sense to wait until it gets to master. Unless, you're more involved in the staging flow.

But I don't think there is any chance at this point that the new Apple SDK and darwin minimum version patterns get removed from `staging-next`, as they were a long time coming before they were merged into staging. And these are the only two things from `staging-next` this PR relies on. Also @emilazy is already aware of the `NIX_CFLAGS_COMPILE` workaround in this PR for the currently disabled `-isysroot`, so she knows where to find and remove it once #349555 progresses.

**Comment by GaetanLepage** [comment]
> > Good job on this @niklaskorz ! Do you want to merge this against `staging-next` or wait that it lands on `master` first and then merge it on `master` ?
> 
> I am fine with both. On the "Nix on macOS" Matrix room I asked what the usual procedure was and @khaneliman answered:
> 
> > If there's a dependency on something in staging/staging-next I don't think it's necessarily wrong to target them if you don't want to wait until it propagates down. I think there's a chance stuff gets ripped out or reverted in those branches, though so it might make sense to wait until it gets to master. Unless, you're more involved in the staging flow.
> 
> But I don't think there is any chance at this point that the new Apple SDK and darwin minimum version hook get removed from `staging-next`, as they were a long time coming before they were merged into staging. And these are the only two things from `staging-next` this PR relies on. Also @emilazy is already aware of the `NIX_CFLAGS_COMPILE` workaround in this PR for the currently disabled `-isysroot`, so she knows where to find and remove it once #349555 progresses.

Ok, let's go for it then :)

**Review by GaetanLepage — APPROVED** [review]

File: pkgs/by-name/ze/zed-editor/package.nix (hunk 2 above)
Comment (jansol) [reply pkgs/by-name/ze/zed-editor/package.nix]: Yeah, hence not an actual review :D  It just caught my eye on the above screenshots

**Review by jansol — APPROVED** [review]

**Review by tymscar — APPROVED** [review]

**Pull request merged**